sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -434,179 +465,156 @@ class GpuDynamicPartitionDataSingleWriter(
+              }
+              None
+            } else {
+              splits(partIx) = null
+              Some(partContigTable)
             }
-            currentWriterStatus.outputWriter.writeAndClose(batch, statsTrackers)
-          } else {
-            writeTableAndClose(concat, outDataTypes,  maxRecordsPerFile, partPath)
           }
-        } else {
-          if (!needSplitBatch(
-            maxRecordsPerFile,
-            currentWriterStatus.recordsInFile,
-            table.getRowCount()))
-          {
-            val batch = GpuColumnVector.from(table, outDataTypes)
-            closeOnExcept(batch) { _ =>
-              statsTrackers.foreach(_.newBatch(currentWriterStatus.outputWriter.path(), batch))
-              currentWriterStatus.recordsInFile += batch.numRows()
-            }
-            currentWriterStatus.outputWriter.writeAndClose(batch, statsTrackers)
-          } else {
-            writeTableAndClose(table, outDataTypes, maxRecordsPerFile, partPath)
+        }
+        contigTableToWrite.foreach { ct =>
+          // we need to split it according to needSplitBatch
+          val maxRecordsPerFileSplits = withResource(ct) { _ =>
+            splitToFitMaxRecords(
+              partContigTable.getTable,
+              maxRecordsPerFile,
+              currentWriterStatus.recordsInFile)
           }
+          writeTableSplitsAndClose(
+            maxRecordsPerFileSplits, outDataTypes, maxRecordsPerFile, partPath)
         }
-      })
-    } finally {
-      if (needToCloseBatch) {
-        cb.close()
-      }
-
-      if (partitionColumns != null) {
-        partitionColumns.close()
-      }
-
-      if (distinctKeys != null) {
-        distinctKeys.close()
       }
+    }
+  }
 
-      if (outputColumns != null) {
-        outputColumns.close()
+  private def updateCurrentWriterIfNeeded(
+    partPath: String,
+    cachesMap: Option[mutable.HashMap[String, WriterStatusWithCaches]]):
+  Option[WriterStatusWithCaches] = {
+    var savedStatus: Option[WriterStatusWithCaches] = None
+    if (currentPartPath != partPath) {
+      val previousPartPath = currentPartPath
+      currentPartPath = partPath
+
+      // see a new partition, close the old writer
+      val previousWriterStatus = currentWriterStatus
+      if (previousWriterStatus != null) {
+        releaseWriter(previousWriterStatus.outputWriter)
       }
 
-      splits.safeClose()
-
-      if (cbKeys != null) {
-        cbKeys.close()
+      if (cachesMap.isDefined) {
+        savedStatus = cachesMap.get.get(currentPartPath)
+        if (savedStatus.isDefined) {
+          // first try to restore the saved writer status,
+          // `GpuDynamicPartitionDataConcurrentWriter` may already opened the writer, and may
+          // have pending caches
+          currentWriterStatus = savedStatus.get.writerStatus
+          // entire batch that is sorted, see a new partition, the old write status is useless
+          cachesMap.get.remove(previousPartPath)
```

Comment:
```suggestion
        // entire batch that is sorted, see a new partition, the old write status is useless
        savedStatus = cachesMap.get.remove(currentPartPath)
        if (savedStatus.isDefined) {
          // first try to restore the saved writer status,
          // `GpuDynamicPartitionDataConcurrentWriter` may already opened the writer, and may
          // have pending caches
          currentWriterStatus = savedStatus.get.writerStatus
```

Reply:
oh wait, I can't do that. The code checks whether `currentPartPath` exists, but then removes `previousPartPath`.